src/stan/mcmc/hmc/integrators/impl_leapfrog.hpp

```diff
@@ -0,0 +1,113 @@
+#ifndef STAN_MCMC_HMC_INTEGRATORS_IMPL_LEAPFROG_HPP
+#define STAN_MCMC_HMC_INTEGRATORS_IMPL_LEAPFROG_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/mcmc/hmc/integrators/base_leapfrog.hpp>
+
+namespace stan {
+  namespace mcmc {
+
+    template <typename Hamiltonian>
+    class impl_leapfrog: public base_leapfrog<Hamiltonian> {
+    public:
+      impl_leapfrog(): base_leapfrog<Hamiltonian>(),
+                       max_num_fixed_point_(10),
+                       fixed_point_threshold_(1e-8) {}
+
+      void begin_update_p(
+        typename Hamiltonian::PointType& z,
+        Hamiltonian& hamiltonian,
+        double epsilon,
+        interface_callbacks::writer::base_writer& info_writer,
+        interface_callbacks::writer::base_writer& error_writer) {
+        hat_phi(z, hamiltonian, epsilon, info_writer, error_writer);
+        hat_tau(z, hamiltonian, epsilon, this->max_num_fixed_point_,
+                info_writer, error_writer);
+      }
+
+      void update_q(typename Hamiltonian::PointType& z,
+                    Hamiltonian& hamiltonian,
+                    double epsilon,
+                    interface_callbacks::writer::base_writer& info_writer,
+                    interface_callbacks::writer::base_writer& error_writer) {
+        // hat{T} = dT/dp * d/dq
+        Eigen::VectorXd q_init = z.q + 0.5 * epsilon * hamiltonian.dtau_dp(z);
+        Eigen::VectorXd delta_q(z.q.size());
+
+        for (int n = 0; n < this->max_num_fixed_point_; ++n) {
+          delta_q = z.q;
+          z.q.noalias() = q_init + 0.5 * epsilon * hamiltonian.dtau_dp(z);
+          hamiltonian.update_metric(z, info_writer, error_writer);
+
+          delta_q -= z.q;
+          if (delta_q.cwiseAbs().maxCoeff() < this->fixed_point_threshold_)
```

Comment:
(comment) This could be a bit more efficient as a loop that breaks when it finds a coeff greater than the threshold.